File: doublelist/java/DoubleList.java

/**
 * DoubleList
 */
public class DoubleList {

    private Node head;
    private Node tail;

    public void addFromTail(Object val) {
        Node n = new Node(val);
        if(this.head == null){
            this.head = n;
            this.tail = n;
        }else{
            this.tail.next = n;
            n.prev = this.tail;
            this.tail = n;
        }
    }

    public void addFromHead(Object val){
        Node n = new Node(val);
        if(this.head == null){
            this.head = n;
            this.tail = n;
        }else{
            n.next = this.head;
            this.head.prev = n;
            this.head = n;
        }
    }


    public void traverse() {
        Node n = this.head;
        while (n != null) {
            System.out.println(n.val);
            n = n.next;
        }
    }

    public void reverse(){
        Node n = this.tail;
        while(n != null){
            System.out.println(n.val);
            n = n.prev;
        }
    }

    public boolean contains(Object val) {
        Node n = this.head;
        while(n != null){
            if(n.val == val){
                return true;
            }
            n = n.next;
        }
        return false;
    }

    public boolean remove(Object val){
        if(this.head == null)return false;
        
        if(this.head.val == val){
            if(this.head == this.tail){
                this.head = null;
                this.tail = null;
            }else{
                this.head = this.head.next;
                this.head.prev = null;
            }
            return true;
        }
        Node n = this.head.next;
        while (n != null) {
            if(n.val == val){
                if(n == this.tail){
                    this.tail = this.tail.prev;
                    this.tail.next = null;
                }else{
                    n.prev.next = n.next;
                    n.next.prev = n.prev;
                }
                
                return true;
            }
            n = n.next;
        }
        return false;
    }


    public static void main(String[] args) {
        DoubleList dl = new DoubleList();
        for(int i = 0;i<10;i++){
            dl.addFromTail(i);
        }
        dl.addFromHead(32);
        dl.addFromHead(4);
        dl.traverse();
        System.out.println();
        //dl.reverse();
        dl.remove(9);
        dl.traverse();
    }
}

File: doublelist/go/DoubleList.go
package main

import (
	"fmt"
)

type Node struct {
	val  interface{}
	prev *Node
	next *Node
}

type DoubleList struct {
	head *Node
	tail *Node
}

func (d *DoubleList) AddFromTail(val interface{}) {
	n := &Node{val: val, prev: nil, next: nil}
	if d.head == nil {
		d.head = n
		d.tail = n
	} else {
		d.tail.next = n
		n.prev = d.tail
		d.tail = n
	}
}

func (d *DoubleList) AddFromHead(val interface{}) {
	n := &Node{val: val, prev: nil, next: nil}
	if d.head == nil {
		d.head = n
		d.tail = n
	} else {
		d.head.prev = n
		n.next = d.head
		d.head = n
	}
}

func (d *DoubleList) Traverse() {
	n := d.head
	for n != nil {
		fmt.Println(n.val)
		n = n.next
	}
}

func (d *DoubleList) Reverse() {
	n := d.tail
	for n != nil {
		fmt.Println(n.val)
		n = n.prev
	}
}

func (d *DoubleList) Contains(val interface{}) bool {
	n := d.head
	for n != nil {
		if n.val == val {
			return true
		}
		n = n.next
	}

	return false
}

func (d *DoubleList) Remove(val interface{}) bool {
	if d.head == nil {
		return false

	}

	if d.head.val == val {
		if d.head == d.tail {
			d.head = nil
			d.tail = nil
		} else {
			d.head = d.head.next
			d.head.prev = nil
		}

		return true
	}

	n := d.head.next
	for n != nil {
		if n.val == val {
			if n == d.tail {
				d.tail = d.tail.prev
				d.tail.next = nil
			} else {
				n.prev.next = n.next
				n.next.prev = n.prev
			}
			return true
		}
		n = n.next
	}
	return false
}

func main() {
	dl := DoubleList{}
	for i := 0; i < 10; i++ {
		dl.AddFromTail(i)
	}
	dl.AddFromHead(32)
	dl.AddFromHead(56)
	dl.Traverse()
	fmt.Println()
	
	fmt.Println(dl.Remove(32))
	dl.Reverse()
}
